envoy/buffer/buffer.h

```diff
@@ -479,6 +504,23 @@ class Instance {
    */
   virtual bool highWatermarkTriggered() const PURE;
 
+protected:
+  /**
+   * Reserve contiguous memory segment with specific size. The first address of this memory segment
+   * will be returned.
+   * @param expected contiguous memory segment size.
+   * @return the first address of this memory segment.
+   */
+  virtual uint8_t* inlineReserve(size_t size) PURE;
```

Comment:
cc @ggreenway 

Why can't you use the existing reserveSingleSlice API?

Reply:
`reserveSingleSlice` is great to reserve a continuous memory segment. But every time we call `addFragments`, we call `inlineReserve` once to get a few tens of bytes of memory. `reserveSingleSlice` seems too heavy for reserve this short continuous memory. 

In the previous version, I have implemented a helper class which will reserve a long continuous memory (4K+) from buffer. And when the local reservation is overflow then the helper class will commit this reservation to the buffer. In that implementation `reserveSingleSlice` is chosen.

Reply:
reserveSingleSlice has a mode where it can append to the previous segment, see separate_slice default argument.  reserveSingleSlice will allocate multiples of 4k

I think the behavior of reserveSingleSlice and inlineReserve is effectively the same.

Reply:
> I think the behavior of reserveSingleSlice and inlineReserve is effectively the same.

The `reserveSingleSlice` and related `commit` is designed for the third-party use and there would be some external overhead for security. But only benchmark can tell us the result. I can implement a temporary version to test it. If there is no performance difference, we can replace inlineReserve with reserveSingleSlice.

Reply:
I implement a new method name `addFragmentsForTest` which use reserveSingleSlice to reserve continuous memory.

```
template <class... Args> size_t Instance::addFragmentsForTest(Args&&... args) {
  size_t total_size_to_write = 0;
  if constexpr (sizeof...(args) > 0) {
    constexpr auto WriteHelper = [](uint8_t** dst, absl::string_view v) {
      memcpy(*dst, v.data(), v.size()); // NOLINT(safe-memcpy)
      *dst += v.size();
    };

    total_size_to_write = (absl::string_view(args).size() + ...);
    auto dst = reserveSingleSlice(total_size_to_write);
    auto* dst_memory_to_write = static_cast<uint8_t*>(dst.slice().mem_);
    (WriteHelper(&dst_memory_to_write, absl::string_view(args)), ...);
    dst.commit(total_size_to_write);
  }
  return total_size_to_write;
}
```

Here are some benchmark results. As you can see, the gap between the two is very obvious. 

```
addFragmentsvsaddFragmentsForTest/0/67108864/1/1     740467928 ns    740479926 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/16/1     49204484 ns     49209840 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/64/1     20073809 ns     20077669 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/128/1    16858928 ns     16861918 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/4096/1   13148293 ns     13151384 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/1/5     330394641 ns    330417087 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/16/5     27311422 ns     27314433 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/64/5     18379040 ns     18381836 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/128/5    15739106 ns     15741881 ns            1
addFragmentsvsaddFragmentsForTest/0/67108864/4096/5   10598302 ns     10601211 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/1/1    4488233887 ns   4484025941 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/16/1    286104634 ns    286123848 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/64/1     71852908 ns     71859333 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/128/1    42036012 ns     42039947 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/4096/1   13553761 ns     13556600 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/1/5    1139331691 ns   1139373311 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/16/5     75401239 ns     75407566 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/64/5     30698843 ns     30702341 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/128/5    22723027 ns     22725917 ns            1
addFragmentsvsaddFragmentsForTest/1/67108864/4096/5   11438161 ns     11440617 ns            1

```

Reply:
@antoniovicente  Thanks for you review. I have posted the implementation which use `reserveSingleSlice` in the last reply.

```
// Test api for `reserveSingleSlice`.
template <class... Args> size_t Instance::addFragmentsForTest(Args&&... args) {
  size_t total_size_to_write = 0;
  if constexpr (sizeof...(args) > 0) {
    constexpr auto WriteHelper = [](uint8_t** dst, absl::string_view v) {
      memcpy(*dst, v.data(), v.size()); // NOLINT(safe-memcpy)
      *dst += v.size();
    };

    total_size_to_write = (absl::string_view(args).size() + ...);
    auto dst = reserveSingleSlice(total_size_to_write);
    auto* dst_memory_to_write = static_cast<uint8_t*>(dst.slice().mem_);
    (WriteHelper(&dst_memory_to_write, absl::string_view(args)), ...);
    dst.commit(total_size_to_write);
  }
  return total_size_to_write;
}
```